Refactor dumpgenerator into subpackages & Remove some useless files

NyaMisty opened the pull request:
## Why?

1. Python packages should only contain Python code and necessary resources, instead of everything
2. Sources files should organized into multiple-level directories, instead in a flat way like this.
   -  This will cause confusion: e.g. logs.py are actually dumping Special:Logs instead of doing logging
   -  This will coupling everything together: devs cannot easily know which python file belongs to which feature, slowing down development.

## What did I do?

- Move non-PyPI package files in `wikiteam3` into resources or tools
  - `listofwikis` -> `resources`
  - `batchdownload` and other single python files except `gui.py` `launcher.py` `uploader.py` -> `tools`
- Refactor into multi-level subpackage
  - `wikiteam3.dumpgenerator`: dumpgenerator top-level package
    - `api`: code related to mediawiki api requesting
    - `cli`: code related to console output & interact
    - `dump`: code for real dumping
      - `xmlrev`: xml revision dump feature
      - `page`: page-based dump feature
      - `misc`: other metadata dumping
    - `log`: helpers for logging
  - `wikiteam3.utils`: various non-dumpgenerator specific functions, so that `launcher` and `uploader` won't need to import `dumpgenerator` module

NyaMisty committed:
Move non-package files to outside folder (resources or tools) & Remove non-repo folder (tests, .vscode)

NyaMisty committed:
Refactor dumpgenerator into sub-packages

robkam commented:
This needs discussing.

NyaMisty commented:
Yep. I'm prepared for discussion :)

randomnetcat commented:
Moving the list of wikis into `resources` seems harmless and obviously fine, can you split that into a separate PR?

elsiehupp commented:
In the grand scheme of things, I think it would be fine to remove the various extraneous upstream WikiTeam bits and bobs from the main branch of this repository. They can always be brought back, and regardless they will still be available on the upstream WikiTeam repository.

NyaMisty commented:
> Moving the list of wikis into `resources` seems harmless and obviously fine, can you split that into a separate PR?

@randomnetcat Hi I believe those simple move are not in a hurry. I can extract it into a extra PR if this PR closed.

Also I did the refactor using PyCharm, which is pretty stable for project without dynamic imports.

This PR is only a base of various other refactors. Based on this PR I'm continuing refactoring loadConfig from giving dict into a dataclass, and then introduce more linting and unit test. You can see them in saveweb/wikiteam3 repo's config_refactor and add_test branch

elsiehupp commented:
Hi @NyaMisty! I just added you to the organization, which should help with collaboration. You can see my notes [here](https://github.com/mediawiki-client-tools/wikiteam3/discussions/61) and [here](https://github.com/mediawiki-client-tools/wikiteam3/discussions/76).

I definitely welcome more substantial refactoring... if you want to take a look at the [`prepare-for-publication` branch](https://github.com/mediawiki-client-tools/wikiteam3/tree/prepare-for-publication), that was where I was working on refactoring before I got overwhelmed.

(Obviously you can take whatever initiative you think is appropriate; I'm just mentioning this if you wanted to compare my approach. The main thing I made the most progress on with that branch was getting the tests to pass, but I got a bit sidetracked along the way. Also I tried to get things set up with type checking using mypy, but I don't remember how far along I got with that.)

robkam commented:
#77

NyaMisty commented:
> Hi @NyaMisty! I just added you to the organization, which should help with collaboration. You can see my notes [here](https://github.com/mediawiki-client-tools/wikiteam3/discussions/61) and [here](https://github.com/mediawiki-client-tools/wikiteam3/discussions/76).
> 
> I definitely welcome more substantial refactoring... if you want to take a look at the [`prepare-for-publication` branch](https://github.com/mediawiki-client-tools/wikiteam3/tree/prepare-for-publication), that was where I was working on refactoring before I got overwhelmed.
> 
> (Obviously you can take whatever initiative you think is appropriate; I'm just mentioning this if you wanted to compare my approach. The main thing I made the most progress on with that branch was getting the tests to pass, but I got a bit sidetracked along the way. Also I tried to get things set up with type checking using mypy, but I don't remember how far along I got with that.)

I've generally reviewed your code. Let me summarize them:
1. changing everything to dump_generator
2. changing everything from camelCase to snake_case
3. spliting more functions
4. many gui development
5. adding config's dict type hint and some other
6. some test case added.

And I've got some advices:
1. among these changes, 1,2,3 causes Git fails to detect renaming, which will be horrible for blaming :(. I suggest we keep the existing camelCase.
2. for GUI that's pretty excellent work, but wikiteam usually runs headlessly on server, so we actually should have GUI. But maybe we should have a better, more responsive CLI
3. Config and Requests are actually the main barrier to typehinting according to my research. config['XXX'] appears everywhere so mypy and PyCharm linters gets overwhelmed. Now I've refactored them.
4. For testing, tox is too outdated, and we should actually use pytests or nose now (which pytests are far more friendly). In unit test we are still discussing whether to use mocked response (can custom more strange & buggy response) or uses a real server (easy to implement).
